src/for.jl

```diff
@@ -0,0 +1,97 @@
+using Distributions
+import Distributions.logpdf
+using Base.Cartesian
+
+
+export logpdf
+export rand
+
+export For
+struct For{F,T,D,X} 
+    f :: F  
+    θ :: T
+end
+
+#########################################################
+# T <: NTuple{N,J} where {J <: Integer}
+#########################################################
+
+For(f, θ::J...) where {J <: Integer} = For(f,θ)
+
+function For(f::F, θ::T) where {F, N, J <: Integer, T <: NTuple{N,J}}
+    d = f.(ones(Int, N)...)
+    D = typeof(d)
+    X = eltype(d)
+    For{F, NTuple{N,J}, D, X}(f,θ)
+end
+
+@inline function logpdf(d::For{F,T,D,X1},xs::AbstractArray{X2,N}) where {F, N, J <: Integer, T <: NTuple{N,J}, D,  X1,  X2 <: X1}
+    s = 0.0
+    @inbounds @simd for θ in CartesianIndices(d.θ)
+        s += logpdf(d.f(Tuple(θ)...), xs[θ])
```

Comment:
This is not ideal for Tracker and GPU support. The better version here is:
```julia
g(cind) = f(Tuple(cind)...)
inds = CartesianIndices(d.θ)
s = sum(logpdf.(g.(inds), xs))
```
This will be slower on the CPU and when using `ForwardDiff` but for `Tracker` it will be more efficient.

```julia
function testf1(x)
    f(i, j) = Normal(i + j)
    θ = (10, 10)
    g(cind) = f(Tuple(cind)...)
    inds = CartesianIndices(θ)
    s = sum(logpdf.(g.(inds), x))
    return s
end

function testf2(x)
    f(i, j) = Normal(i + j)
    _θ = (10, 10)
    s = 0.0
    @inbounds @simd for θ in CartesianIndices(_θ)
        s += logpdf(f(Tuple(θ)...), x[θ])
    end
    return s
end
x = rand(10, 10)
xparam = param(x)

using BenchmarkTools

@btime testf1($x); # 714 ns
@btime testf2($x); # 634 ns

@btime testf1($xparam); # 1.25 μs
@btime testf2($xparam); #14.6 μs
```

Reply:
Moderate hit on floats in exchange for great AD benefits, works for me. IIUC `Tracker` is here for now but will go away as Zygote matures, any thoughts how this will effect the comparison?

Also, I've played with some options for doing this with a parallel reduction, i.e. thread-local on a CPU. Maybe this should be a different combinator, maybe `ParFor`? Or as a new argument could be ok too, whatever is best for users without sacrificing performance

Reply:
Zygote promises to support loops like this efficiently, but in my experience Zygote is still not stable enough to adopt in a relatively mature project like Turing. The parallel stuff in Julia 1.3 also looks very interesting. I still haven't experimented with it though, so I will have to do that before commenting any further.

Reply:
For `Tracker` support, could it make more sense to just specify the adjoint separately?

There's some discussion of parallel reduction in this context here:
https://discourse.julialang.org/t/parallel-reductions/30180/8

Reply:
I mean it's possible but getting it to perform better than the baseline implementation with broadcasting might not be easy.